File: ReformatGWAScols.py
#! usr/bin/env python3
# -*- encoding: utf-8 -*-
'''
@File    :    ReformatGWAScols.py
@Time    :    2025/01/09 17:41:09
@Author  :    Lulu Shi
@Mails   :    crazzy_rabbit@163.com
@line    :    https://github.com/Crazzy-Rabbit
'''
import click
import pandas as pd

@click.command()
@click.option('--gwas', type=str, help="GWAS summary data", required=True)
@click.option('--SNP', type=str, help="SNP colnames", required=True)
@click.option('--A1', type=str, help="A1 colnames for effect allele", required=True)
@click.option('--A2', type=str, help="A2 colnames for another allele", required=True)
@click.option('--freq', type=str, help="frequency colnames for effect allele", required=True)
@click.option('--beta', type=str, help="beta colnames for effect allele", required=True)
@click.option('--se', type=str, help="se colnames", required=True)
@click.option('--pval', type=str, help="p colnames", required=True)
@click.option('--num', type=str, help="sample size colnames or sample size number", required=True)
@click.option('--outprefix', type=str, help="outprefix for your file", required=True)
def process_gwas(gwas, SNP, A1, A2, freq, beta, se, pval, num, outprefix):
    '''
    reformat gwas colnames and select cols that used in downstream analysis
    '''
    # Load GWAS summary data
    gwas_dt = pd.read_csv(gwas, delimiter='\t')

    # Check if num is a numeric value or column name
    if num.isdigit():
        gwas_dt['N'] = int(num)
        columns_map = {SNP: 'SNP', A1: 'A1', A2: 'A2', freq: 'freq', beta: 'beta', se: 'se', pval: 'p', 'N': 'N'}
    elif num in gwas_dt.columns:
        columns_map = {SNP: 'SNP', A1: 'A1', A2: 'A2', freq: 'freq', beta: 'beta', se: 'se', pval: 'p', num: 'N'}
    else:
        raise ValueError(f"Column {num} neither numeric nor found in the data")

    # Rename columns
    gwas_dt.rename(columns=columns_map, inplace=True)
    gwas_dt.to_csv(outprefix, sep='\t', index=False)

if __name__ == '__main__':
    process_gwas()

File: ReformatGWAScols.r
library(data.table)
library(optparse)
option_list <- list(
    make_option(c("--gwas", type="character"), help="GWAS summary data"),
    make_option(c("--SNP"), type="character", help="SNP colnames"),
    make_option(c("--A1"), type="character", help="A1 colnames for effect allele"),
    make_option(c("--A2"), type="character", help="A1 colnames for another allele"),
    make_option(c("--freq"), type="character", help="frequency colnames for effect allele"),
    make_option(c("--beta"), type="character", help="beta colnames for effect allele"),
    make_option(c("--se"), type="character", help="se colnames"),
    make_option(c("--pval"), type="character", help="p colnames"),
    make_option(c("--num"), type="character", help="sample size colnames or sample size number"),
    make_option(c("--outprefix"), type="character", help="outprefix for your file")
)
opt_parser <- OptionParser(option_list=option_list)
opts <- parse_args(opt_parser)

gwas <- opts$gwas
SNP  <- opts$SNP
A1   <- opts$A1
A2   <- opts$A2
freq <- opts$freq
beta <- opts$beta
se   <- opts$se
pval <- opts$pval
num  <- opts$num

# set col names
gwas_dt <- fread(gwas)
if (!is.na(as.numeric(num)) && !grepl("[a-zA-Z]", num)){
    gwas_dt$N <- as.numeric(num)
    old_col <- c(SNP, A1, A2, freq, beta, se, pval, "N")
    new_col <- c("SNP", "A1", "A2", "freq", "beta", "se", "p", "N")
    setnames(gwas_dt, old_col, new_col)
} else {
    if (num %in% colnames(gwas_dt)) {
        old_col <- c(SNP, A1, A2, freq, beta, se, pval, num)
        new_col <- c("SNP", "A1", "A2", "freq", "beta", "se", "p", "N")
        setnames(gwas_dt, old_col, new_col)
    } else {
        stop(paste("Column", num, "neither numeric nor found in the data"))
    }
}

fwrite(gwas_dt, file=outprefix, sep="\t")
